File: python/24.py
with open('input/24.txt') as f:
    blocks = [x.strip().split('\n') for x in f.read().strip().split('inp')[1:]]

digits = []
for block in blocks:
    up = block[4].endswith('1')
    sub = int(block[5].split(' ')[-1])
    val = int(block[-3].split()[-1])
    digits.append((up, sub, val))

def solve(vars, inc, res=0):
    if len(vars) == 0 and res == 0:
        return ''
    up, sub, val = vars[0]
    if up:
        for i in list(range(1,10))[::inc]:
            if (rest := solve(vars[1:], inc, val + i + 26*res)) is not None:
                return str(i) + rest
    else:
        if not 1 <= (digit := (res % 26) + sub) <= 9:
            return None
        if (rest := solve(vars[1:], inc, res // 26)) is not None:
            return str(digit) + rest
    return None 

print("Part 1:", solve(digits,-1))
print("Part 2:", solve(digits, 1))

File: c/24.c
#include "common.h"

int data[14][3];

void parse() {
  FILE *f = fopen(INFILE("24.txt"), "r");
  char buf[2048];

  int block = -1;
  int line = 0;
  while (GETLINE(buf, f)) {
    if (strncmp(buf, "inp", 3) == 0) {
      block++;
      line = 0;
    } else {
      line++;
      if (line == 4) {
        data[block][0] = buf[strlen(buf) - 1] == '1';
      } else if (line == 5) {
        assert(sscanf(buf, "add x %d", &data[block][1]) == 1);
      } else if (line == 15) {
        assert(sscanf(buf, "add y %d", &data[block][2]) == 1);
      }
    }
  }
  fclose(f);
}

bool solve(int idx, bool big, int z, char result[15]) {
  if (idx == 14 && z == 0)
    return true;
  if (data[idx][0]) {
    for (int i = 1; i < 10; i++) {
      int j = big ? 10 - i : i;
      int new_z = 26*z + data[idx][2] + j;
      if (solve(idx+1, big, new_z, result)) {
        result[idx] = j + '0';
        return true;
      }
    }
  } else {
    int digit = (z % 26) + data[idx][1];
    if (1 <= digit && digit <= 9 && solve(idx+1, big, z/26, result)) {
      result[idx] = digit + '0';
      return true;
    }
  }
  return false;
}

int main() {
  parse();
  char result1[15] = {0};
  char result2[15] = {0};
  solve(0, true, 0, result1);
  solve(0, false, 0, result2);
  printf("Part 1: %s\n", result1);
  printf("Part 2: %s\n", result2);
  return 0;
}